### PR TITLE
add zipWith & zipWithIndex with BiFunction

### PR DESCRIPTION
**Pull request by ronanM**
Added on Traversable:
- ```<U, R> Traversable<R> zipWith(Iterable<? extends U> that, BiFunction<? super T, ? super U, ? extends R> mapper);```
- ```<U> Traversable<U> zipWithIndex(BiFunction<? super T, ? super Long, ? extends U> mapper);```

See #1313

### CODE REVIEW

File: javaslang/src/test/java/javaslang/collection/IntMap.java [hunk 1]
```diff
@@ -23,7 +23,7 @@
     @SuppressWarnings("unchecked")
     public static <T> IntMap<T> of(Map<Integer, T> original) {
         return original.isEmpty() ? (IntMap<T>) EMPTY
-                                  : new IntMap<>(original);
+                : new IntMap<>(original);
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/IntMap.java]: Minor: please leave the original indentation in these cases.
In Idea you can configure it in:
```
File | Settings | Editor | Code Style | Java
Wrapping and Braces | Ternary operation | Align when multiline | ☑
```

File: javaslang/src/test/java/javaslang/collection/IntMultimap.java [hunk 2]
```diff
@@ -313,8 +313,14 @@ public int characteristics() {
 
     @Override
     public <U> Seq<Tuple2<T, U>> zip(Iterable<? extends U> that) {
+        return zipWith(that, Tuple::of);
+    }
+
+    @Override
+    public <U, R> Seq<R> zipWith(Iterable<? extends U> that, BiFunction<? super T, ? super U, ? extends R> mapper) {
         Objects.requireNonNull(that, "that is null");
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/IntMultimap.java]: `that is null` doesn't sound very useful to me.
Could we rename `that` to `target` or something less context dependent :)?

File: javaslang/src/main/java/javaslang/collection/Traversable.java [hunk 3]
```diff
@@ -1203,4 +1219,19 @@ default Number sum() {
      * @return A new traversable containing all elements of this traversable paired with their index, starting with 0.
      */
     Traversable<Tuple2<T, Long>> zipWithIndex();
+
+    /**
+     * Returns a traversable formed from this traversable and another Iterable collection by mapping elements.
+     * If one of the two iterables is longer than the other, its remaining elements are ignored.
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Traversable.java]: @danieldietrich is it ok that we're ignoring the leftovers?
Shouldn't they have the same length?

File: javaslang/src/main/java/javaslang/collection/Traversable.java [hunk 4]
```diff
@@ -1203,4 +1219,19 @@ default Number sum() {
...
+     * <p>
+     * The length of the returned traversable is the minimum of the lengths of this traversable and {@code that}
+     * iterable.
+     *
+     * @param <U>    The type of the mapped elements.
+     * @param mapper a mapper.
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Traversable.java]: please provide a more useful description :)

File: javaslang/src/main/java/javaslang/collection/PriorityQueue.java [hunk 5]
```diff
@@ -558,8 +571,8 @@ private Node(T root, int rank, Seq<Node<T>> children) {
                 assert rank == tree.rank;
 
                 return comparator.isLessOrEqual(this.root, tree.root)
-                       ? of(this.root, this.rank + 1, tree.appendTo(this.children))
-                       : of(tree.root, tree.rank + 1, this.appendTo(tree.children));
+                        ? of(this.root, this.rank + 1, tree.appendTo(this.children))
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/PriorityQueue.java]: Minor: unnecessary formatting (the previous was better aligned)

File: javaslang/src/test/java/javaslang/collection/IntMap.java (hunk 1 above)
Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/IntMap.java]: We provide a formatter for IntelliJ, see `.ide/idea-settings.jar`. Do you use it?

File: javaslang/src/main/java/javaslang/collection/LinkedHashSet.java [hunk 6]
```diff
@@ -846,7 +852,13 @@ public String mkString(CharSequence prefix, CharSequence delimiter, CharSequence
 
     @Override
     public LinkedHashSet<Tuple2<T, Long>> zipWithIndex() {
-        return LinkedHashSet.ofAll(iterator().zipWithIndex());
+        return zipWithIndex(Tuple::of);
+    }
+
+    @Override
+    public <U> LinkedHashSet<U> zipWithIndex(BiFunction<? super T, ? super Long, ? extends U> mapper) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/LinkedHashSet.java]: Can we weaken the return type?

**Comment by danieldietrich** [comment]
Very nice! I will check the changes now...

**Comment by danieldietrich** [comment]
Very nice! I will check the changes now...

File: javaslang/src/test/java/javaslang/collection/IntMap.java (hunk 1 above)
Comment (ronanM) [inline javaslang/src/test/java/javaslang/collection/IntMap.java]: Yes, I've imported settings.

Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/IntMap.java]: Than please format only the `VCS changed` lines :) (or manually revert the ones that were better formatted before)

File: javaslang/src/test/java/javaslang/collection/IntMultimap.java (hunk 2 above)
Comment (ronanM) [inline javaslang/src/test/java/javaslang/collection/IntMultimap.java]: All other `zip` function have `that` parameter named.

Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/IntMultimap.java]: I know.

File: javaslang/src/main/java/javaslang/collection/Traversable.java (hunk 4 above)
Comment (ronanM) [inline javaslang/src/main/java/javaslang/collection/Traversable.java]: I follow `map` and `flatMap`.

File: javaslang/src/main/java/javaslang/collection/Queue.java [hunk 7]
```diff
@@ -1135,11 +1135,17 @@ public boolean startsWith(Iterable<? extends T> that, int offset) {
...
+        return zipWith(that, Tuple::of);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <U, R> Queue<R> zipWith(Iterable<? extends U> that, BiFunction<? super T, ? super U, ? extends R> mapper) {
         Objects.requireNonNull(that, "that is null");
-        return ofAll(toList().zip((Iterable<U>) that));
+        Objects.requireNonNull(mapper, "mapper is null");
+        return ofAll(toList().zipWith((Iterable<U>) that, mapper));
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Queue.java]: Does a type-hint work instead of casting? `return ofAll(toList().<U> zipWith(that, mapper));`

File: javaslang/src/test/java/javaslang/collection/IntMultimap.java (hunk 2 above)
Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/IntMultimap.java]: Let's leave it as is. We have `that` in many places and is correct to say `this.zipWith(that)`.

By convention be write `<param-name> is null`. @ronanM has done it correct.

File: javaslang/src/main/java/javaslang/collection/Queue.java (hunk 7 above)
Comment (ronanM) [inline javaslang/src/main/java/javaslang/collection/Queue.java]: With type-hints it doesn't compile. Strange !?!

File: javaslang/src/main/java/javaslang/collection/Traversable.java (hunk 4 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Traversable.java]: We could fix the descriptions of map, flatMap, zipWithIndex et.al. in a separate PR

File: javaslang/src/main/java/javaslang/collection/Traversable.java (hunk 3 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Traversable.java]: `zipWithIndex` has no other Iterable, so the description is not correct. After zipWithIndex we have the same length.

File: javaslang/src/main/java/javaslang/collection/Queue.java (hunk 7 above)
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Queue.java]: You can try extract to variable (ctrl alt v) make it compile and inline it back (ctrl alt v).
This usually works for me :)

File: javaslang/src/main/java/javaslang/collection/PriorityQueue.java (hunk 5 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/PriorityQueue.java]: Looks like number of spaces % 4 is 0. It is likely the IDE changed indentation. For me it is ok...

File: javaslang/src/main/java/javaslang/collection/LinkedHashSet.java (hunk 6 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/LinkedHashSet.java]: Why? It is correct in my opinion.

File: javaslang/src/test/java/javaslang/collection/IntMap.java (hunk 1 above)
Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/IntMap.java]: It is my fault. We should provide a formatter that aligns rows correct when applied to a file. There most by some formatting option to do that. We have to fix the formatter and check it in (in a different PR). We can then also reformat the whole code base (modulo Generator.scala and generated code).

@ronanM has done all right. You may leave it as is!

File: javaslang/src/main/java/javaslang/collection/Queue.java (hunk 7 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Queue.java]: It was just a question. I think you are right and it does not compile then. Just leave it as is.

File: javaslang/src/main/java/javaslang/collection/PriorityQueue.java (hunk 5 above)
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/PriorityQueue.java]: It's the same ternary alignment I mentioned below.
It's an unrelated modification anyway (that doesn't make the code better), it's reason I recommended a revert.

File: javaslang/src/main/java/javaslang/collection/LinkedHashSet.java (hunk 6 above)
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/LinkedHashSet.java]: It's like returning an `LinkedList`, the user of the code shouldn't care whether we change it to `ArrayList`, they should only see `List` in the public interface.
The same applies here, we should probably return `Set`, so that we can freely modify the underlying structure later.

File: javaslang/src/main/java/javaslang/collection/Queue.java (hunk 7 above)
Comment (ronanM) [inline javaslang/src/main/java/javaslang/collection/Queue.java]: I just push the deletion of this cast.

File: javaslang/src/main/java/javaslang/collection/LinkedHashSet.java (hunk 6 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/LinkedHashSet.java]: I've thought about it - and took a look at the Scala API reference. For example Scala's [TreeSet.zipWithIndex](http://www.scala-lang.org/api/2.11.8/index.html#scala.collection.immutable.TreeSet@zipWithIndex:scala.collection.immutable.TreeSet[(A,Int)]) returns `TreeSet`.
For now I would keep it as is. I see currently no reason to change it.

**Comment by danieldietrich** [comment]
The PR LGTM. I will pull it now. If I should have forgotten s.th. @paplorinc remarked (other than formatting which we will cover in another PR) please give me a hint.

**Comment by danieldietrich** [comment]
Btw - many thanks, great work!
